FLT/Deformations/RepresentationTheory/Basic.lean

```diff
@@ -0,0 +1,379 @@
+  letI := f.toAlgebra
+  haveI : ContinuousSMul A B := continuousSMul_of_algebraMap A B hf
+  rw [GaloisRep.frame_baseChange]
+  rfl
+
+lemma FramedGaloisRep.baseChange_map [IsTopologicalRing B]
+    (ρ : FramedGaloisRep K A n) (f : A →+* B) (hf : Continuous f)
+    (g : K →+* L) : (ρ.baseChange f hf).map g = (ρ.map g).baseChange f hf := rfl
+
+lemma Matrix.map_trace {F α β n : Type*} [AddCommMonoid β] [AddCommMonoid α] [Fintype n]
+    (M : Matrix n n α) (f : F) [FunLike F α β] [AddMonoidHomClass F α β] :
+    (M.map f).trace = f M.trace :=
+  (AddMonoidHom.map_trace f M).symm
+
+lemma Matrix.map_det {F α β n : Type*} [CommRing β] [CommRing α] [Fintype n]
+    [DecidableEq n]
+    (M : Matrix n n α) (f : F) [FunLike F α β] [RingHomClass F α β] :
+    (M.map f).det = f M.det :=
+  (RingHom.map_det (f : α →+* β) M).symm
+
+@[simp]
+lemma LinearMap.trace_toLin' {R n : Type*} [CommSemiring R] [DecidableEq n]
+    [Fintype n] (M : Matrix n n R) : LinearMap.trace _ _ M.toLin' = M.trace := by
+  rw [LinearMap.trace_eq_matrix_trace _ (Pi.basisFun ..), LinearMap.toMatrix_eq_toMatrix',
+      LinearMap.toMatrix'_toLin']
+
+omit [NumberField K] in
+lemma FramedGaloisRep.det_baseChange [IsTopologicalRing B]
+    (ρ : FramedGaloisRep K A n) (f : A →+* B) (hf : Continuous f) :
+    (ρ.baseChange f hf).det = .comp ⟨f, hf⟩ ρ.det := by
+  ext σ
+  dsimp [baseChange, GaloisRep.det]
+  rw [GL_symm_apply]
+  dsimp [← Matrix.toLin'_apply']
+  rw [LinearMap.det_toLin', Matrix.map_det, LinearMap.det_toMatrix']
+
+/-- Given a (global) galois rep, this is the local galois rep at a finite prime `v`.
+Note: this fixes an arbitrary embedding `Kᵃˡᵍ → Kᵥᵃˡᵍ`, or equivalently,
+an arbitrary choice of valuation on `Kᵃˡᵍ` extending `v`. -/
+noncomputable
+abbrev GaloisRep.adic (ρ : GaloisRep K A M) (v : Ω K) : GaloisRep (v.adicCompletion K) A M :=
+  ρ.map (algebraMap _ _)
+
+universe v u
+variable {R : Type u} [CommRing R]
+
+/-- The class of galois reps unramified at `v`. -/
+class GaloisRep.IsUnramifiedAt (ρ : GaloisRep K A M) (v : Ω K) : Prop where
+  localInertiaGroup_le :
+    letI := moduleTopology A (Module.End A M)
+    localInertiaGroup v ≤ (ρ.adic v).ker
+
+instance (ρ : GaloisRep K A M) (v : Ω K) [ρ.IsUnramifiedAt v] (e : M ≃ₗ[A] N) :
+    (ρ.conj e).IsUnramifiedAt v where
+  localInertiaGroup_le := (GaloisRep.IsUnramifiedAt.localInertiaGroup_le (ρ := ρ)).trans (by simp)
+
+instance [IsTopologicalRing B] [Algebra A B] [ContinuousSMul A B]
+    [Module.Finite A M] [Module.Free A M] (ρ : GaloisRep K A M) (v : Ω K) [ρ.IsUnramifiedAt v] :
+    (ρ.baseChange B).IsUnramifiedAt v :=
+  ⟨(GaloisRep.IsUnramifiedAt.localInertiaGroup_le (ρ := ρ)).trans
+    (((ρ.adic v).ker_baseChange (B := B)))⟩
+
+variable [Module.Free A M] [Module.Finite A M] [Module.Free A N] [Module.Finite A N]
+
+/-- The character polynomial of the frobenious conjugacy class at `v` under `ρ`. -/
+noncomputable
+def GaloisRep.charFrob (ρ : GaloisRep K A M) : Polynomial A := (ρ.adic v Frobᵥ).charpoly
```

Comment:
So is your vision that Frob_v should be an element of a local Galois group rather than a global one? Should we have a predicate `IsFrobeniusAt v` on global Galois groups?